docs/observability-developer-guide.md

```diff
@@ -0,0 +1,202 @@
+objects, and show up the objects they relate to when running `kubectl describe`.
+
+Events should be recorded in the following cases:
+
+* A significant operation is taken on a resource
+* The state of a resource is changed
+* An error occurs
+
+The events recorded in these cases can be thought of as forming an event log of
+things that happen for the resources that Crossplane manages. Each event should
+refer back to the relevant controller and resource, and use other fields of the
+Event kind as appropriate.
+
+More details about examples of how to interact with events can be found in the
+guide to [debugging an application cluster].
+
+## Choosing between methods of error reporting
+
+There are many ways to report errors, such as:
+
+* Metrics
+* Events
+* Logging
+* Tracing
+
+It can be confusing to figure out which one is appropriate in a given situation.
+This section will try to offer advice and a mindset that can be used to help
+make this decision.
+
+Let's set the context by listing the different user scenarios where error
+reporting may be consumed. Here are the typical scenarios as we imagine them:
+
+1. A person **using** a system needs to figure out why things aren't working as
+   expected, and whether they made a mistake that they can correct.
+2. A person **operating** a service needs to monitor the service's **health**,
+   both now and historically.
+3. A person **debugging** a problem which happened in a **live environment**
+   (often an **operator** of the system) needs information to figure out what
+   happened.
+4. A person **developing** the software wants to **observe** what is happening.
+5. A person **debugging** the software in a **development environment**
+   (typically a **developer** of the system) wants to debug a problem (there is
+   a lot of overlap between this and the live environment debugging scenario).
+
+The goal is to satisfy the users in all of the scenarios. We'll refer to the
+scenarios by number.
+
+The short version is: we should do whatever satisfies all of the scenarios.
+Logging and events are the recommendations for satisfying the scenarios,
+although they don't cover scenario 2.
+
+The longer version is:
+
+* Scenario 1 is best served by events in the context of Crossplane, since the
+  users may not have access to read logs or metrics, and even if they did, it
+  would be hard to relate them back to the event the user is trying to
+  understand.
+* Scenario 2 is best served by metrics, because they can be aggregated and
+  understood as a whole. And because they can be used to track things over time.
+* Scenario 3 is best served by either logging that contains all the information
+  about and leading up to the event. Or, if available, some sort of error event
+  reporting system which captures all of the relevant information in one place
+  (for example, a well-configured project in [Sentry]). In
+  my experience, the larger a fleet gets, the less useful logging is, and the
```

Comment:
nit: switches to first-person here, which feels a little weird with the rest of this doc and others

Reply:
I decided to remove this part about Sentry; I considered doing so earlier given that we don't use an event aggregator like Sentry and (AFAIK) it is uncommon for OSS to do so.